[Bug] [Master]  State event handle error, get a unknown exception, will retry this event with IndexOutOfBoundsException

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.


### What happened

```shell
[ERROR] 2023-06-16 11:50:22.653 org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable:[304] - [WorkflowInstance-239515][TaskInstance-null] - State event handle error, get a unknown exception, will retry this event: WorkflowStateEvent(processInstanceId=239515, taskInstanceId=null, status=WorkflowExecutionStatus{code=6, desc='failure'}, type=PROCESS_STATE_CHANGE, key=null, channel=null, context=null)
java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:659)
	at java.util.ArrayList.get(ArrayList.java:435)
	at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable.processComplementData(WorkflowExecuteRunnable.java:609)
	at org.apache.dolphinscheduler.server.master.event.WorkflowStateEventHandler.handleStateEvent(WorkflowStateEventHandler.java:56)
	at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable.handleEvents(WorkflowExecuteRunnable.java:289)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

version 3.1.4 

`src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java` L608-L610

```java
        if (scheduleDate == null) {
            scheduleDate = complementListDate.get(0); // if complementListDate is null , IndexOutOfBoundsException
        } else if (processInstance.getState().isFinished()) {
```



### What you expected to happen

No error message is sent, and the workflow ends normally.

### How to reproduce

complete data.

### Anything else

_No response_

### Version

3.1.x

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java
+++ b/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java
@@ -672,7 +672,7 @@ public class WorkflowExecuteRunnable implements Callable<WorkflowSubmitStatus> {
 
         Date scheduleDate = processInstance.getScheduleTime();
         if (scheduleDate == null) {
-            if (complementListDate.size() == 0) {
+            if (CollectionUtils.isEmpty(complementListDate)) {
                 logger.info("complementListDate size is zero, process complement end. process id:{}", processInstance.getId());
                 return true;
             }

--- a/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java
+++ b/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java
@@ -673,7 +673,8 @@ public class WorkflowExecuteRunnable implements Callable<WorkflowSubmitStatus> {
         Date scheduleDate = processInstance.getScheduleTime();
         if (scheduleDate == null) {
             if (CollectionUtils.isEmpty(complementListDate)) {
-                logger.info("complementListDate size is zero, process complement end. process id:{}", processInstance.getId());
+                log.info("complementListDate is empty, process complement end. process id:{}", processInstance.getId());
+
                 return true;
             }
             scheduleDate = complementListDate.get(0);

--- a/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java
+++ b/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java
@@ -672,6 +672,10 @@ public class WorkflowExecuteRunnable implements Callable<WorkflowSubmitStatus> {
 
         Date scheduleDate = processInstance.getScheduleTime();
         if (scheduleDate == null) {
+            if (complementListDate.size() == 0) {
+                logger.info("complementListDate size is zero, process complement end. process id:{}", processInstance.getId());
+                return true;
+            }
             scheduleDate = complementListDate.get(0);
         } else if (processInstance.getState().isFinished()) {
             endProcess();

